iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,327 @@
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Actor for Peer {
+    async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        //self.addr = Some(ctx.addr());
+        // Subscribe reading stream
+        ctx.notify_with(self.stream());
+    }
+}
+
+#[async_trait::async_trait]
+impl Handler<MessageResult> for Peer {
+    type Result = ();
+
+    async fn handle(&mut self, msg: MessageResult) {
+        match msg {
+            MessageResult::Ok { message } => {
+                // We got handshake or the real message from other peer
+                match message {
+                    Message::ClientHello { key} => {
+                        let public_key: PublicKey = PublicKey::from(key);
+                        let shared = self.secret_key.diffie_hellman(&public_key);
+                        let mut nonce = [0u8; 12];
+                        {
+                            let mut rng = rand::thread_rng();
+                            rng.fill_bytes(&mut nonce);
+                        }
+                        let chacha = Chacha::new(shared.as_bytes(), &nonce);
+                        self.cipher = Some(chacha);
+                        let mut write_half = self.write.as_mut();
+                        send_server_hello(&mut write_half.unwrap(), self.public_key.as_bytes(), &nonce).await;
+                        self.state = State::ServerHelloSent;
+                    },
+                    Message::ServerHello { key, nonce } => {
+                        let public_key: PublicKey = PublicKey::from(key);
+                        let shared = self.secret_key.diffie_hellman(&public_key);
+                        let chacha = Chacha::new(shared.as_bytes(), &nonce);
+                        self.cipher = Some(chacha);
+                        self.state = State::Ready;
+                    },
+                    Message::Data { data } => {
+                        let data = match &self.cipher {
+                            None => { data }
+                            Some(cipher) => {
+                                match cipher.decrypt(data.as_slice()) {
+                                    Ok(data) => data,
+                                    Err(e) => {
+                                        warn!("Error decrypting message! {}", e);
+                                        self.state = State::Error;
+                                        return;
+                                    }
+                                }
+                            }
+                        };
+                        let msg = ReceivedPeerMsg { data, id: self.id.clone() };
+                        self.network_addr.do_send(msg).await;
+                    },
+                    Message::Error => {
+                        self.state = State::Error;
+                    }
+                }
+
+            }
+            MessageResult::Err { error } => {
+                warn!("Error in peer read! {}", error);
```

Comment:
You could move out nested match  out of this match. That would be way more readable

Reply:
Done.

Reply:
Not done... Outer branch is not linear still.

Reply:
I mean match on `MessageResult`